src/app/modules/service-providers/services/navigation/navigation.service.ts

```diff
@@ -1,16 +1,101 @@
+import { select, Store } from '@ngrx/store';
+import { lineSlice, lineString, nearestPointOnLine, point, length } from '@turf/turf';
+import {
+  currentPositionSelector, getNavState,
+  nextWaypointIndexSelector,
+  routeDetailsSelector,
+  routePointsSelector
+} from '../../../../store/nav.selectors';
+import { filter, tap, withLatestFrom } from 'rxjs/operators';
+import {
+  PlayVoiceWpNotificationAction,
+  SetDistanceToEndpointAction,
+  SetIsNextWpNotifiedAction,
+  SetNextWaypointDistanceAction,
+  SetNextWaypointIndexAction
+} from '../../../../store/nav.actions';
+
 
 @Injectable({
   providedIn: 'root'
 })
 export class NavigationService {
 
-  constructor(private store: Store<any>) { }
+  constructor(private store: Store<any>) {
+    this.calcNextWaypoints();
+  }
+
+
+  calcNextWaypoints() {
+    this.store.pipe(
+      select(currentPositionSelector),
+      withLatestFrom(this.store.pipe(select(routePointsSelector))),
+      filter(([currentPosition, routePoints]) => currentPosition.latitude !== null && routePoints !== null),
+      withLatestFrom(this.store.pipe(select(nextWaypointIndexSelector)), this.store.pipe(select(routeDetailsSelector))),
+      tap(([[currentPosition, routePoints], nextWpIndex, routeDetails]) => {
+        const currentPositionOnRoute = nearestPointOnLine(lineString(routePoints), point([currentPosition.latitude,
+          currentPosition.longitude
+        ]));
+        this.store.dispatch(new SetDistanceToEndpointAction({
+          distanceToEndpoint:
+            (routeDetails.routeLength - (currentPositionOnRoute.properties.location * 1000))
+        }));
+        let newNextWpIndex = nextWpIndex;
+        while (newNextWpIndex < routeDetails.routeLegs.length - 1 &&
+        routeDetails.routeLegs[newNextWpIndex].index <= currentPositionOnRoute.properties.index) {
+          newNextWpIndex++;
+        }
+        if (newNextWpIndex > nextWpIndex) {
+          this.store.dispatch(new SetNextWaypointIndexAction({ nextWaypointIndex: newNextWpIndex }));
+        }
 
+        const lineToNextWp = lineSlice(point([currentPositionOnRoute.geometry.coordinates[0],
+            currentPositionOnRoute.geometry.coordinates[1]
+          ]),
+          point(routeDetails.routeLegs[newNextWpIndex].coords),
+          lineString(routePoints));
+        const distanceToNextWp = length(lineToNextWp) * 1000;
+        this.store.dispatch(new SetNextWaypointDistanceAction({ nextWaypointDistance: distanceToNextWp }));
+      }),
+      withLatestFrom(this.store.pipe(select(getNavState))),
+      filter(([[[currentPosition, routePoints], nextWpIndex, routeDetails], stateData]) => stateData.previousPosition !== null &&
+        !stateData.isNextWpNotified && stateData.nextWaypointDistance !== null),
+      tap(([[[currentPosition, routePoints], nextWpIndex, routeDetails], stateData]) => {
+        const currentPositionOnRoute = nearestPointOnLine(lineString(routePoints), point([currentPosition.latitude,
```

Comment:
routePoints are "static" and can be kept as lineString instead of re-creating it on every update.
This snap to road is already calculated in this flow, it should be mapped out instead of re-calculate.
It can also be stored to avoid also calculating the previous point